src/nyc_trees/apps/home/training/types.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+from __future__ import print_function
+from __future__ import unicode_literals
+from __future__ import division
+
+from django.core.urlresolvers import reverse
+from django.contrib.auth.decorators import login_required
+from django_tinsel.decorators import route
+from django_tinsel.utils import decorate as do
+from apps.home.training.decorators import render_flatpage, mark_user
+from django.contrib.flatpages.views import flatpage
+
+_PURE_TEMPLATE = '%s_pure'
+_MARK_PROGRESS_TEMPLATE = '%s_mark_progress'
+_TRAINING_FINISHED_TEMPLATE = 'training_finished_%s'
+
+
+class TrainingGateway(object):
+    def __init__(self, view, first_step, last_step):
+        self.previous_step = last_step
+        self.next_step = first_step
+        self.view = view
+        self.name = None
+
+    def is_flatpage(self):
+        return False
+
+    def get_step(self, name):
+        for step in self.get_steps():
+            if step.name == name:
+                return step
+        raise ValueError()
+
+    def is_complete(self, user):
+        return all([step.is_complete(user) for step in self.get_steps()])
+
+    def get_context(self, user):
+        last_was_complete = True
+        next_step = None
+
+        training_steps = []
+        for step in self.get_steps():
+            is_complete = step.is_complete(user)
+            if next_step is None and last_was_complete and not is_complete:
+                next_step = step
+                is_next = True
+            else:
+                is_next = False
+            training_steps.append({'step': step,
+                                   'is_complete': is_complete,
+                                   'is_next': is_next})
+            last_was_complete = is_complete
+
+        return training_steps
+
+    def get_steps(self):
+        steps = []
+        step = self.next_step
+        while step and step != self:
+            steps.append(step)
+            step = step.next_step
+        return steps
+
+    def pure_kwargs(self):
+        return {'name': '%s_pure' % self.name,
+                'view': login_required(self.view)}
+
+
+class TrainingStep(object):
```

Comment:
Sounds like a good idea, because then `is_flatpage` can then be a typecheck, ie, `isinstance(step, FlatpageTrainingStep)`